streams: Panic on collecting failed records

I started to see persistent panics like the below in my production deployment:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x7fda4c8ce9b1]

goroutine 83 [running]:
github.com/s12v/awsbeats/streams.collectFailedEvents(0xc420ad1dd0, 0xc420f4b300, 0x32, 0x3b4, 0xc420ad1dd0, 0x0, 0x0)
        /go/src/github.com/s12v/awsbeats/streams/client.go:170 +0xd1
github.com/s12v/awsbeats/streams.(*client).publishEvents(0xc4200da930, 0xc420f4b300, 0x32, 0x3b4, 0x1d14c30, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/s12v/awsbeats/streams/client.go:92 +0x2be
github.com/s12v/awsbeats/streams.(*client).Publish(0xc4200da930, 0x14b9860, 0xc422b27940, 0xc42008ce40, 0xc420f69f78)
        /go/src/github.com/s12v/awsbeats/streams/client.go:64 +0x45
github.com/elastic/beats/libbeat/outputs.(*backoffClient).Publish(0xc42000d8a0, 0x14b9860, 0xc422b27940, 0x0, 0x0)
        /go/src/github.com/elastic/beats/libbeat/outputs/backoff.go:43 +0x4b
github.com/elastic/beats/libbeat/publisher/pipeline.(*netClientWorker).run(0xc4200af280)
        /go/src/github.com/elastic/beats/libbeat/publisher/pipeline/output.go:90 +0x1a9
created by github.com/elastic/beats/libbeat/publisher/pipeline.makeClientWorker
        /go/src/github.com/elastic/beats/libbeat/publisher/pipeline/output.go:31 +0xf0
```

This indicates that we're receiving records with unexpected structures from kinesis streams. I have no reference to the concrete specification of kinesis records that I'm unable to "fix" it.

In the meantime, I'd make awsbeats gracefully degrade, that is to give up retrying failed records with unexpected structures but just leave some log messsage to help further investigation.

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,7 +4,6 @@ go:
   - "1.10"
 
 install:
-  - go get github.com/elastic/beats || true
   - curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
   - dep ensure
 

--- a/Makefile
+++ b/Makefile
@@ -14,7 +14,7 @@ BEAT_DOCKER_IMAGE ?= docker.elastic.co/beats/$(BEAT_NAME):$(BEATS_VERSION)
 GOPATH ?= $(HOME)/go
 
 .PHONY: all
-all: vars test beats build
+all: vars beats test build
 
 .PHONY: vars
 vars:

--- a/streams/client.go
+++ b/streams/client.go
@@ -167,6 +167,10 @@ func collectFailedEvents(res *kinesis.PutRecordsOutput, events []publisher.Event
 				logp.Warn("no record returned from kinesis for event: ", events[i])
 				continue
 			}
+			if r.ErrorCode == nil {
+				logp.Warn("skipping failed event with unexpected state: corresponding kinesis record misses error code: ", r)
+				continue
+			}
 			if *r.ErrorCode != "" {
 				failedEvents = append(failedEvents, events[i])
 			}

--- a/streams/client_test.go
+++ b/streams/client_test.go
@@ -181,6 +181,28 @@ func TestPublishEvents(t *testing.T) {
 	}
 
 	{
+		// Records with nil error codes should be ignored with some log
+		client.encoder = StubCodec{dat: []byte("boom"), err: nil}
+		client.streams = StubClient{
+			out: &kinesis.PutRecordsOutput{
+				Records: []*kinesis.PutRecordsResultEntry{
+					&kinesis.PutRecordsResultEntry{
+						ErrorCode: nil,
+					},
+				},
+				FailedRecordCount: aws.Int64(1),
+			},
+		}
+		rest, err := client.publishEvents(events)
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+		if len(rest) != 0 {
+			t.Errorf("unexpected number of remaining events: %d", len(rest))
+		}
+	}
+
+	{
 		// Kinesis received the event but it was not persisted, probably due to underlying infrastructure failure
 		client.encoder = StubCodec{dat: []byte("boom"), err: nil}
 		client.streams = StubClient{

